### PR TITLE
Use CMake's C# support to build DacTableGen instead of manually invoking csc.exe ourselves.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
CMake 3.8 added support for compiling C# projects against desktop csc.exe. In our build scripts, we use the desktop csc.exe to build the DacTableGen tool (which we later use on CoreCLR.dll). Since we use CMake 3.11 or newer on Windows (see #23868), the CMake version we use has support for these types of targets. This allows us to have the DacTableGen project be "just another project" in our native build and for us to more easily/cleanly express the dependencies between the coreclr target and the DacTableGen target to CMake.

The change from the `-DFEATURE_X=1` to `-DFEATURE_X` is required because CMake passes all defines along to the C# compiler as well, and the defines with values don't work in C#. Additionally, a `-DFEATURE_X` is equivalent to `-DFEATURE_X=1` in all C++ compilers we use.

### CODE REVIEW

**Comment by hoyosjs** [comment]
Changes functionally look okay, save for the macro errors...

File: src/ToolBox/SOS/DacTableGen/CMakeLists.txt [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
-# (2) This CSC is invoking the desktop .NET Framework CSC, not the buildtools version.
-add_custom_target(dactablegen ALL 
-    COMMAND csc.exe /t:exe /platform:anycpu32bitpreferred /r:System.dll /r:DiaLib.dll /nowarn:1668 /out:${CMAKE_CURRENT_BINARY_DIR}/dactablegen.exe ${DACTABLEGEN_SOURCES}
-    COMMAND ${CMAKE_COMMAND} -E copy DIAlib.dll ${CMAKE_CURRENT_BINARY_DIR}
-    DEPENDS ${DACTABLEGEN_SOURCES} DIAlib.dll
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-)
+# Remove defines that aren't just by-name but actually have defined values.
+# C# doesn't support preprocessor defines with values.
+remove_definitions(-D_SECURE_SCL=0 -DWINVER=0x0602 -D_WIN32_WINNT=0x0602)
```
Comment (janvorli) [inline src/ToolBox/SOS/DacTableGen/CMakeLists.txt]: Since the dactablegen build doesn't need any of the defines that were defined for C++, it seems it would be simpler to just clear the DIRECTORY_COMPILE_DEFINITIONS instead of changing all of the definitions that were defined as values and then massaging them using the regex.

File: src/ToolBox/SOS/DacTableGen/CMakeLists.txt [hunk 2]
```diff
@@ -1,23 +1,20 @@
...
-#   warning CS1668: Invalid search path 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.14.26428\atlmfc\lib\ARM64' specified in 'LIB environment variable' -- 'The system cannot find the path specified. '
-# There are several problems here: (1) the C++ project files are adding non-existent directories to the LIB path, especially for ARM and ARM64,
-# (2) This CSC is invoking the desktop .NET Framework CSC, not the buildtools version.
-add_custom_target(dactablegen ALL 
-    COMMAND csc.exe /t:exe /platform:anycpu32bitpreferred /r:System.dll /r:DiaLib.dll /nowarn:1668 /out:${CMAKE_CURRENT_BINARY_DIR}/dactablegen.exe ${DACTABLEGEN_SOURCES}
-    COMMAND ${CMAKE_COMMAND} -E copy DIAlib.dll ${CMAKE_CURRENT_BINARY_DIR}
-    DEPENDS ${DACTABLEGEN_SOURCES} DIAlib.dll
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-)
+# Remove defines that aren't just by-name but actually have defined values.
```
Comment (janvorli) [inline src/ToolBox/SOS/DacTableGen/CMakeLists.txt]: The comment now doesn't match the reality.

File: src/pal/src/libunwind/src/CMakeLists.txt [hunk 3]
```diff
@@ -15,7 +15,7 @@ endif()
 # The HAVE___THREAD set to 1 causes creation of thread local variable with tls_model("initial-exec")
 # which is incompatible with usage of the unwind code in a shared library.
 add_definitions(-DHAVE___THREAD=0)
-add_definitions(-DHAVE_CONFIG_H=1)
+add_definitions(-DHAVE_CONFIG_H)
```
Comment (janvorli) [inline src/pal/src/libunwind/src/CMakeLists.txt]: This is a setting for a third party library whose source we include. Can you please revert the change here?

File: src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 4]
```diff
@@ -149,41 +149,45 @@ endif(FEATURE_EVENT_TRACE)
...
-        COMMENT Add dactable, debug resources, and dump helper resources to coreclr
-    )
-
-    if(NOT DEFINED CLR_CROSS_COMPONENTS_BUILD)
-        add_custom_target(inject_debug_resources ALL DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/inject_debug_resources.timestamp)
+      # Add dac table & debug resource to coreclr
+      get_include_directories(INC_DIR)
+      get_compile_definitions(PREPROCESS_DEFINITIONS)
+      list(APPEND INC_DIR -I${CLR_DIR}/src/vm -I${CLR_DIR}/src/vm/${ARCH_SOURCES_DIR} -I${CLR_DIR}/src/debug/ee -I${CLR_DIR}/src/gc)
+      list(APPEND PREPROCESS_DEFINITIONS -DDACCESS_COMPILE -DDBG_TARGET_64BIT=1 -DDBG_TARGET_WIN64)
```
Comment (janvorli) [inline src/dlls/mscoree/coreclr/CMakeLists.txt]: There is a "-DDBG_TARGET_64BIT=1" that you've missed on this line.

**Comment by jkoritzinsky** [comment]
Looks like AzDO is having network issues...

**Comment by jkoritzinsky** [comment]
@hoyosjs I've fix the issue with the dac table resource embedding target not running. I messed up my generator regex and also didn't correctly chain in the "fail on build" version of the target.

**Comment by hoyosjs** [comment]
I'd like to make e2e testing, but I verified that the resources are now embedded in the image.

**Comment by jkoritzinsky** [comment]
I've verified that I can debug through an MVC application with a runtime built in this branch.

Once the AzDO run completes, I'll merge this in.

**Comment by jkoritzinsky** [comment]
Test timeout is unrelated to this PR, so I'm going to merge.
